### PR TITLE
Implement the hypercenter and upper central series for finite dimensional Lie algebras

### PR DESCRIPTION
**Pull request by tscrim**
We implement the upper central series and hypercenter of a finite dimensional Lie algebra.

As part of the testing, we found improvements could be made to the generic example (needed so the example would work with these tests): We moved the `reduce()` method up to the category. Implemented methods expected for the subalgebras. Better input verification. Elements are now hashable. We had to work around the fact that `lift()` is serving multiple roles as to the UEA and to the ambient space for a Lie subalgebra. This functionality needs to be divided (I take full responsibility for this bad design) by renaming `lift` to something for the UEA map, but this is a significant change that should be done on a followup PR.

We also fixed some bugs and improved efficiency with changing vectors into Lie algebra elements.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/algebras/lie_algebras/quotient.py [hunk 1]
```diff
@@ -291,11 +294,26 @@ def _repr_(self):
             L: General linear Lie algebra of rank 2 over Rational Field
             I: Ideal ([0 0]
             [0 1])
+
+            sage: L = LieAlgebras(QQ).FiniteDimensional().WithBasis().example()
+            sage: a,b,c = L.gens()
+            sage: I = L.ideal([a+2*b, b+3*c])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/quotient.py]:
```suggestion
            sage: I = L.ideal([a + 2*b, b + 3*c])
```

File: src/sage/categories/examples/finite_dimensional_lie_algebras_with_basis.py [hunk 2]
```diff
@@ -135,6 +139,57 @@ def _element_constructor_(self, x):
...
+        EXAMPLES::
+
+            sage: L = LieAlgebras(QQ).FiniteDimensional().WithBasis().example()
+            sage: a, b, c = L.gens()
+            sage: L.lift(a)
+            b0
+            sage: L.lift(b).parent() is L.universal_enveloping_algebra()
+            True
+
+            sage: I = L.ideal([a+2*b, b+3*c])
```
Comment (mkoeppe) [inline src/sage/categories/examples/finite_dimensional_lie_algebras_with_basis.py]:
```suggestion
            sage: I = L.ideal([a + 2*b, b + 3*c])
```

File: src/sage/categories/examples/finite_dimensional_lie_algebras_with_basis.py [hunk 3]
```diff
@@ -302,7 +359,36 @@ def from_vector(self, v, order=None):
...
 
+    def leading_monomials(self):
+        r"""
+        Return the set of leading monomials of the basis of ``self``.
+
+        EXAMPLES::
+
+            sage: L = LieAlgebras(QQ).FiniteDimensional().WithBasis().example()
+            sage: a, b, c = L.lie_algebra_generators()
+            sage: I = L.ideal([2*a+b, b + c])
```
Comment (mkoeppe) [inline src/sage/categories/examples/finite_dimensional_lie_algebras_with_basis.py]:
```suggestion
            sage: I = L.ideal([2*a + b, b + c])
```

File: src/sage/categories/finite_dimensional_lie_algebras_with_basis.py [hunk 4]
```diff
@@ -2068,3 +2166,71 @@ def basis_matrix(self):
...
+                are not the leading supports of the basis of ``self``.
+
+                EXAMPLES:
+
+                An example reduction in a 6 dimensional Lie algebra::
+
+                    sage: sc = {('a','b'): {'d': 1}, ('a','c'): {'e': 1},
+                    ....:       ('b','c'): {'f': 1}}
+                    sage: L.<a,b,c,d,e,f> = LieAlgebra(QQ, sc)
+                    sage: I =  L.ideal(c)
```
Comment (mkoeppe) [inline src/sage/categories/finite_dimensional_lie_algebras_with_basis.py]:
```suggestion
                    sage: I = L.ideal(c)
```

File: src/sage/algebras/lie_algebras/quotient.py [hunk 5]
```diff
@@ -218,8 +218,11 @@ def __classcall_private__(cls, I, ambient=None, names=None,
         index_set = [i for i in sorted_indices if i not in I_supp]
 
         if names is None:
-            amb_names = dict(zip(sorted_indices, ambient.variable_names()))
-            names = [amb_names[i] for i in index_set]
+            try:
+                amb_names = dict(zip(sorted_indices, ambient.variable_names()))
+                names = [amb_names[i] for i in index_set]
+            except ValueError:  # ambient has not assigned variable names
+                pass
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/quotient.py]: I'd suggest to rewrite it as `try... except... else`, which makes it clearer which of the lines is expected to raise the exception

**Review by mkoeppe — APPROVED** [review]
Otherwise LGTM

File: src/sage/algebras/lie_algebras/quotient.py (hunk 5 above)
Comment (tscrim) [reply src/sage/algebras/lie_algebras/quotient.py]: I instead just examined the `_names` attribute and avoided using `try`-`except` to control programming flow.

**Pull request merged**